crates/oxvg_optimiser/src/jobs/inline_styles.rs

```diff
@@ -0,0 +1,1370 @@
+                    iter.for_each(|_| ());
+                    while iter.next_sequence().is_some() {
+                        ids.extend(
+                            iter.filter_map(|t| match t {
+                                selector::Component::ID(id) => Some(id),
+                                _ => None,
+                            })
+                            .map(|t| format!("{t}")),
+                        );
+                    }
+                    ids
+                })
+                .collect(),
+        }
+    }
+}
+
+impl<'de, E: Element> Deserialize<'de> for InlineStyles<E> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let options = Options::deserialize(deserializer)?;
+        Ok(Self {
+            options,
+            styles: RefCell::new(vec![]),
+            removed_tokens: RefCell::new(RemovedTokens::default()),
+            parent_tokens: RefCell::new(ParentTokens::default()),
+        })
+    }
+}
+
+impl Clone for CapturedStyles {
+    fn clone(&self) -> Self {
+        let node = self.node.clone();
+        Self {
+            node,
+            css: self.css.clone(),
+        }
+    }
+}
+
+const DEFAULT_ONLY_MATCHED_ONCE: bool = true;
+const DEFAULT_REMOVE_MATCHED_SELECTORS: bool = true;
+
+lazy_static! {
+    static ref DEFAULT_USE_MQS: Vec<String> = vec![String::new(), String::from("screen")];
+    static ref DEFAULT_USE_PSEUDOS: Vec<String> = vec![String::new()];
+    static ref PRESERVED_PSEUDOS: BTreeSet<&'static str> = {
+        let mut result = PSEUDO_FUNCTIONAL.clone();
+        result.extend(&*PSEUDO_TREE_STRUCTURAL);
+        result
+    };
+}
+
+#[test]
+#[allow(clippy::too_many_lines)]
+fn inline_styles() -> anyhow::Result<()> {
+    use crate::test_config;
+
+    insta::assert_snapshot!(test_config(
+        r#"{ "inlineStyles": {} }"#,
+        Some(
+            r#"<svg id="test" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100">
+    <rect width="100" height="100" class="st0"/>
+</svg>"#
+        ),
+    )?);
+
+    insta::assert_snapshot!(test_config(
+        r#"{ "inlineStyles": {} }"#,
+        Some(
+            r#"<svg id="test" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100">
+    <style>
+        .st0{fill:blue;}
+    </style>
+    <rect width="100" height="100" class="st0"/>
+</svg>"#
+        ),
+    )?);
+
+    insta::assert_snapshot!(test_config(
+        r#"{ "inlineStyles": {} }"#,
+        Some(
+            r#"<svg xmlns="http://www.w3.org/2000/svg" id="dark" viewBox="0 0 258.12 225.88">
+    <!-- for https://github.com/svg/svgo/pull/592#issuecomment-266327016 -->
+    <style>
+        .cls-7 {
+            only-cls-7: 1;
+        }
+        .cls-7,
+        .cls-8 {
+            cls-7-and-8: 1;
+        }
+    </style>
+
+    <path class="cls-7"/>
+</svg>"#
+        ),
+    )?);
+
+    // BUG: Duplicate properties are not removed
```

Comment:
Something wrong with how I'm using lightningcss, we ought to fix this eventually